libraries/AP_GPS/AP_GPS_GSOF.cpp

```diff
@@ -160,46 +183,152 @@ AP_GPS_GSOF::parse(const uint8_t temp)
                           0x00,0x03
-                         }; // checksum
+                         };
 
-    buffer[4] = packetcount++;
+    populateOutgoingTransNumber(buffer);
+    populateChecksum(buffer, sizeof(buffer));
 
-    uint8_t checksum = 0;
-    for (uint8_t a = 1; a < (sizeof(buffer) - 1); a++) {
-        checksum += buffer[a];
-    }
-
-    buffer[17] = checksum;
-
-    port->write((const uint8_t*)buffer, sizeof(buffer));
+    return requestResponse(buffer, sizeof(buffer));
 }
 
-void
+bool
 AP_GPS_GSOF::requestGSOF(const uint8_t messageType, const HW_Port portIndex, const Output_Rate rateHz)
 {
+    Debug("Requesting gsof #%u on port %u", messageType, (uint8_t)portIndex);
+
+    // This packet is not documented in the API.
     uint8_t buffer[21] = {0x02,0x00,0x64,0x0f,0x00,0x00,0x00, // application file record
                           0x03,0x00,0x01,0x00, // file control information block
                           0x07,0x06,0x0a,static_cast<uint8_t>(portIndex),static_cast<uint8_t>(rateHz),0x00,messageType,0x00, // output message record
                           0x00,0x03
-                         }; // checksum
+                         };
+
+    populateOutgoingTransNumber(buffer);
+    populateChecksum(buffer, sizeof(buffer));
+
+    return requestResponse(buffer, sizeof(buffer));
+}
+
+bool
+AP_GPS_GSOF::disableOutput(const HW_Port portIndex) {
+    Debug("Disabling output on on port %u", (uint8_t)portIndex);
+    switch(portIndex) {
+    case HW_Port::COM1:
+    {
+        constexpr uint8_t cmd[8] = {
+            STX,
+            0x00,
+            0x51,
+            0x02,
+            0x0A,
+            0x01,
+            0x5E,
+            ETX
+        };
+        // The checksum is hard-coded.
+        // No need to calculate it.
+        return requestResponse(cmd, sizeof(cmd));
+    }
 
-    buffer[4] = packetcount++;
+    case HW_Port::COM2:
+    {
+        constexpr uint8_t cmd[8] = {
+            STX,
+            0x00,
+            0x51,
+            0x02,
+            0x0A,
+            0x02,
+            0x5F,
+            ETX
+        };
+        return requestResponse(cmd, sizeof(cmd));
+    }
+    }
+    return false;
+}
+
+bool
+AP_GPS_GSOF::requestResponse(const uint8_t* buf, const uint8_t len) {
+    if (!port->is_initialized()) {
+        Debug("Port not initialized");
+        return false;
+    }
+
+    for (int attempt = 0; attempt <= configuration_attempts; attempt++) {
+        // Clear input buffer before doing configuration
+        if (!port->discard_input()) {
+            Debug("Failed to discard input");
+            return false;
+        };
 
+        port->write((const uint8_t*)buf, len);
+
+        constexpr uint16_t expected_bytes = 1;
+        // TODO use wait_timeout instead once HAL_SITL has it implemented.
+        const auto start_wait = AP_HAL::millis();
+        auto now = start_wait;
+        while (now  - start_wait <= configuration_wait_time_ms) {
```

Comment:
This looks like we're sleeping in the main loop for up to 5ms.  That's not acceptable, if so.

Reply:
Yea, either this needs to be on its own thread, or I take the microstrain approach and do the configuration async-style instead of blocking.